spec/predicates/vsa.md

```diff
@@ -1,11 +1,279 @@
+
+> Identifies the entity that performed the verification.
+>
+> The identity MUST reflect the trust base that consumers care about. How
+> detailed to be is a judgment call.
+>
+> Consumers MUST accept only specific (signer, verifier) pairs. For example,
+> "GitHub" can sign provenance for the "GitHub Actions" verifier, and "Google"
+> can sign provenance for the "Google Cloud Deploy" verifier, but "GitHub" cannot
+> sign for the "Google Cloud Deploy" verifier.
+>
+> The field is required, even if it is implicit from the signer, to aid readability and
+> debugging. It is an object to allow additional fields in the future, in case one
+> URI is not sufficient.
+
+<a id="verifier.id"></a>
+`verifier.id` _string ([TypeURI]), required_
+
+> URI indicating the verifier’s identity.
+
+<a id="verifier.version"></a>
+`verifier.version` _map (string->string), optional_
+
+> Map of names of components of the verification platform to their version.
+
+<a id="timeVerified"></a>
+`timeVerified` _string ([Timestamp]), optional_
+
+> Timestamp indicating what time the verification occurred.
+
+<a id="resourceUri"></a>
+`resourceUri` _string ([ResourceURI]), required_
+
+> URI that identifies the resource associated with the artifact being verified.
+
+<a id="policy"></a>
+`policy` _object ([ResourceDescriptor]), required_
+
+> Describes the policy that the `subject` was verified against.
+>
+> The entry MUST contain a `uri`.
+
+<a id="inputAttestations"></a>
+`inputAttestations` _array ([ResourceDescriptor]), optional_
+
+> The collection of attestations that were used to perform verification.
+> Conceptually similar to the `resolvedDependencies` field in [SLSA Provenance].
+>
+> This field MAY be absent if the verifier does not support this feature.
+> If non-empty, this field MUST contain information on _all_ the attestations
+> used to perform verification.
+>
+> Each entry MUST contain a `digest` of the attestation and SHOULD contains a
+> `uri` that can be used to fetch the attestation.
+
+<a id="verificationResult"></a>
+`verificationResult` _string, required_
+
+> Either “PASSED” or “FAILED” to indicate if the artifact passed or failed the policy verification.
+
+<a id="verifiedProperties"></a>
+`verifiedProperties` _array (_string), required_
+
+> Indicates the properties verified for the artifact (and not
+> its dependencies), or "FAILED" if policy verification failed.
+>
+> If including SLSA levels, users MUST NOT include more than one level per SLSA
+> track. Note that each SLSA level implies all levels below it (e.g.
+> SLSA_BUILD_LEVEL_3 implies SLSA_BUILD_LEVEL_2 and SLSA_BUILD_LEVEL_1), so
+> there is no need to include more than one level per track.
+
+<a id="dependencyProperties"></a>
+`dependencyProperties` _object, optional_
+
+> A count of the dependencies verified to have each property.
+>
+> Map from _string to the number of the artifact's _transitive_ dependencies
+> that were verified at the indicated level. Absence of a given value
```

Comment:
If the dependencies are `verified` shouldn't that generate a VSA or some other evidence of its own to reference?

Reply:
Are you asking why we need this map when we could just include all the VSAs for the transitive dependencies?

Reply:
@jkjell You can include VSAs for each transitive dependency in the attestation bundle, but doing so for a nontrivial piece of software will mean creating a large attestation bundle that may be impractical to evaluate. Avoiding that giant bundle is kind of the point of a VSA.

That being said, I'm in favor of removing this field. It isn't clear to me why you'd want dependencies' security properties summarized in the VSA instead of implicitly through the evaluated policy. Alternatively, I could imagine gathering this data on a dashboard if you had a database of VSAs for the subject's transitive dependencies. Such a dashboard could be updated in real time based on whatever data source the VSA producer used to fill these fields. WDYT @TomHennen ?

Reply:
Since VSAs are often used to cross organization boundaries I don't think it's always policy to know the actual policy used across all orgs with all dependencies, likewise it may not be possible to have a real-time dashboard that could provide these feedback cross-org.

Certainly in some cases it may work (seems like a good use case for GUAC), but leaving it in VSAs would solve that problem.

We certainly still have this use case internally so we'd need to support it, maybe by adding an org specific field.  My preference would be to keep it, perhaps as a SLSA specific field addition, but we can probably work around it (I'd just be sad).

> @jkjell You can include VSAs for each transitive dependency in the attestation bundle, but doing so for a nontrivial piece of software will mean creating a large attestation bundle that may be impractical to evaluate. Avoiding that giant bundle is kind of the point of a VSA.

When consuming a piece of software via its VSA you shouldn't need any additional attestations.  So just one VSA.

If you're verifying the provenance of some piece of software though, then you may want VSAs for each of the _direct_ dependencies (which will cover their transitive dependencies).  You don't need them all.  So it shouldn't wind up being that big.

Reply:
> VSAs are often used to cross organization boundaries I don't think it's always policy to know the actual policy used across all orgs with all dependencies

@TomHennen do you mean the policy of the dependency that is verified or the policy of the VSA itself? If policy is unknown for the VSA, then the VSA is nothing more than a signature for external parties. I suppose that might be ok in some situations but, it's a reversion to implicit trust.

Reply:
@kpk47 and I spoke offline, it sounds like a reasonable approach is:

1. Remove `dependencyProperties` from this predicate type (to avoid in-toto needing to define the concept of dependencies and other confusing problems)
2. Have SLSA define something like the `dependencyProperties` field _when using_ the in-toto VSA following [the existing in-toto guidance on extension fields](https://github.com/in-toto/attestation/blob/main/spec/v1/README.md#parsing-rules:~:text=considered%20major%20versions.-,Extension%20fields%3A,-Producers%20MAY%20add).

Does this sound reasonable to everyone?

Reply:
> > VSAs are often used to cross organization boundaries I don't think it's always policy to know the actual policy used across all orgs with all dependencies
> 
> @TomHennen do you mean the policy of the dependency that is verified or the policy of the VSA itself? If policy is unknown for the VSA, then the VSA is nothing more than a signature for external parties. I suppose that might be ok in some situations but, it's a reversion to implicit trust.

I think maybe removing the thing resolves this, but I don't think it's totally true to say that it's "nothing more than a signature".  I agree that it would be ideal if people can see the original evidence and policies that were used throughout the entire supply chain (and the VSA, provenance, and other attestations let people do that), but it won't always be possible (some orgs may not be willing or able to make that data public because of proprietary information).

However, even if they can't make the policy and other attestations public, the VSA still includes the `verifiedProperties` fields and the `verifier.id` field, which means that the VSA is better than a plain signature in one important respect, it makes it clear that the verifier (`verified.id`) is asserting that the software covered by the VSA has some specific but high level properties (`verifiedProperties`).  That can be pretty useful when orgs are exchanging software?

So, the way I see it, plain signature < VSA without public evidence or policy < VSA with public evidence and policy?

But maybe this isn't the place to have that discussion?

Reply:
I think the `verifiedProperties` was the piece I was missing. Establishing trust in a verifier AND that it asserts a given set of properties were confirmed by a policy is a great way of abstracting the value of the policy from the policy itself.